### PR TITLE
Non-catastrophic errors prototype for request exhaustion

### PR DESCRIPTION
**Pull request by yfguo**

### CODE REVIEW

**Comment by yfguo** [comment]
The commit history needs cleanup.

File: configure.ac [hunk 1]
```diff
@@ -602,6 +602,9 @@ AC_ARG_ENABLE(dbg-nolocal, AC_HELP_STRING([--enable-dbg-nolocal], [enables debug
 AC_ARG_ENABLE(dbg-localoddeven, AC_HELP_STRING([--enable-dbg-localoddeven], [enables debugging mode where shared-memory communication is enabled only between even processes or odd processes on a node]),
     AC_DEFINE(ENABLED_ODD_EVEN_CLIQUES, 1, [Define to enable debugging mode where shared-memory communication is done only between even procs or odd procs]))
 
+AC_ARG_ENABLE(runtime-limits, AC_HELP_STRING([--enable-runtime-limits], [enables setting runtime limits to MPICH internal data structures]),
+    AC_DEFINE(ENABLED_RUNTIME_LIMITS, 1, [Define to disable shared-memory communication for debugging]))
```
Comment (hajimefu) [inline configure.ac]: Maybe this description is not correct?

File: src/mpid/ch4/netmod/ofi/ofi_send.h [hunk 2]
```diff
@@ -245,10 +245,10 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send_normal(const void *buf, int count, M
             msg.addr = MPIDI_OFI_comm_to_phys(comm, rank);
 
             MPIDI_OFI_CALL_RETRY(fi_tsendmsg(MPIDI_Global.ctx[0].tx, &msg, flags), tsendv,
-                                                     MPIDI_OFI_CALL_LOCK);
+                                 MPIDI_OFI_CALL_LOCK);
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_send.h]: You can dump the whitespace changes. They'll all come it when #2845 is merged.

File: src/mpi/pt2pt/irecv.c [hunk 3]
```diff
@@ -127,6 +127,8 @@ int MPI_Irecv(void *buf, int count, MPI_Datatype datatype, int source,
 
     mpi_errno = MPID_Irecv(buf, count, datatype, source, tag, comm_ptr,
                            MPIR_CONTEXT_INTRA_PT2PT, &request_ptr);
+    /* If the error is cannot allocate request, goto fail before accessing the request */
+    if (mpi_errno != MPI_SUCCESS) goto fn_fail;
```
Comment (pavanbalaji) [inline src/mpi/pt2pt/irecv.c]: Can you run the `code-cleanup.sh` script on this?

File: src/mpid/ch4/shm/posix/posix_impl.h [hunk 4]
```diff
@@ -106,13 +106,15 @@ typedef struct {
...
         (sreq_) = MPIR_Request_create(MPIR_REQUEST_KIND__SEND); \
+        MPIR_ERR_CHKANDSTMT((sreq_)== NULL, mpi_errno, MPIX_ERR_NO_REQ, goto fn_fail, "**nomemreq"); \
         MPIR_Request_add_ref((sreq_));                          \
         (sreq_)->u.persist.real_request   = NULL;               \
     }
 
 #define MPIDI_POSIX_REQUEST_CREATE_RREQ(rreq_)                  \
     {                                                           \
         (rreq_) = MPIR_Request_create(MPIR_REQUEST_KIND__RECV); \
+        MPIR_ERR_CHKANDSTMT((rreq_)== NULL, mpi_errno, MPIX_ERR_NO_REQ, goto fn_fail, "**nomemreq"); \
```
Comment (pavanbalaji) [inline src/mpid/ch4/shm/posix/posix_impl.h]: I think this should be consistent for all request creations, and not just send and receive requests.

File: src/mpi/pt2pt/irecv.c (hunk 3 above)
Comment (yfguo) [reply src/mpi/pt2pt/irecv.c]: Sure. Will run it once merged with recent changes.

File: src/mpid/ch4/shm/posix/posix_impl.h (hunk 4 above)
Comment (yfguo) [reply src/mpid/ch4/shm/posix/posix_impl.h]: Good point. I will update the PR.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
I have a bad rebase, cleaning it.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

File: src/mpid/ch4/generic/mpidig_send.h [hunk 5]
```diff
@@ -68,7 +71,6 @@ static inline int MPIDI_am_isend(const void *buf, MPI_Aint count, MPI_Datatype d
   fn_exit:
     MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_AM_ISEND);
     return mpi_errno;
-
```
Comment (raffenet) [inline src/mpid/ch4/generic/mpidig_send.h]: unrelated whitespace change

File: src/mpid/ch4/netmod/ucx/ucx_init.h [hunk 6]
```diff
@@ -216,6 +216,7 @@ static inline int MPIDI_NM_mpi_init_hook(int rank,
 
 #ifndef HAVE_DEBUGGER_SUPPORT
     MPIDI_UCX_global.lw_send_req = MPIR_Request_create(MPIR_REQUEST_KIND__SEND);
+    // TODO: is it safe to assume a request will always be allocated
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_init.h]: We should abort if this request is not allocated successfully during init.

File: src/mpid/ch4/generic/mpidig_send.h [hunk 7]
```diff
@@ -90,7 +92,9 @@ static inline int MPIDI_psend_init(const void *buf,
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_PSEND_INIT);
 
     sreq = MPIDI_CH4I_am_request_create(MPIR_REQUEST_KIND__PREQUEST_SEND, 2);
-    *request = sreq;
+    if (sreq == NULL)
+        goto fn_fail;
```
Comment (raffenet) [inline src/mpid/ch4/generic/mpidig_send.h]: no error here?

File: src/mpi/coll/helper_fns.c [hunk 8]
```diff
@@ -39,6 +39,8 @@ int MPIC_Probe(int source, int tag, MPI_Comm comm, MPI_Status * status)
   fn_exit:
     return mpi_errno;
   fn_fail:
+    if (mpi_errno == MPIX_ERR_NOREQ)
+        MPIR_ERR_SET(mpi_errno, MPI_ERR_OTHER, "**nomem");
```
Comment (raffenet) [inline src/mpi/coll/helper_fns.c]: This is weird. Are we resetting the error class to `MPI_ERR_OTHER` here?

File: src/mpid/ch4/netmod/ucx/ucx_init.h (hunk 6 above)
Comment (yfguo) [reply src/mpid/ch4/netmod/ucx/ucx_init.h]: Agree.

File: src/mpid/ch4/generic/mpidig_send.h (hunk 7 above)
Comment (yfguo) [reply src/mpid/ch4/generic/mpidig_send.h]: The outer function that calls this will do the check. Since we only return the address, there is no way to return an error code at here.

File: src/mpi/coll/helper_fns.c (hunk 8 above)
Comment (yfguo) [reply src/mpi/coll/helper_fns.c]: For the collective helper functions, I just convert the non-catastrophic error to a catastrophic one here.

File: src/mpid/ch4/generic/mpidig_send.h (hunk 7 above)
Comment (yfguo) [reply src/mpid/ch4/generic/mpidig_send.h]: Oh, I see what you meant.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

File: src/mpid/ch4/netmod/ucx/ucx_probe.h [hunk 9]
```diff
@@ -33,6 +33,7 @@ static inline int MPIDI_NM_mpi_improbe(int source,
     if (message_h) {
         *flag = 1;
         req = (MPIR_Request *) MPIR_Request_create(MPIR_REQUEST_KIND__MPROBE);
+        MPIR_ERR_CHKANDSTMT((req) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
         MPIR_Assert(req);
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_probe.h]: assert can be removed

File: src/mpid/ch4/netmod/ofi/ofi_impl.h [hunk 10]
```diff
@@ -59,6 +59,7 @@
 #define MPIDI_OFI_request_create_null_rreq(rreq_, mpi_errno_, FAIL_) \
   do {                                                                  \
     (rreq_) = MPIR_Request_create(MPIR_REQUEST_KIND__RECV);             \
+    MPIR_ERR_CHKANDSTMT((rreq_) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq"); \
     if ((rreq_) != NULL) {                                              \
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: Can this branch be removed now?

File: src/mpid/ch4/src/ch4r_rma.h [hunk 11]
```diff
@@ -642,6 +642,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_CH4U_mpi_rput(const void *origin_addr,
     if (sreq == NULL) {
         /* create a completed request for user if issuing is completed immediately. */
         sreq = MPIR_Request_create(MPIR_REQUEST_KIND__RMA);
+        MPIR_ERR_CHKANDSTMT((sreq) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
         MPIR_Assert(sreq);
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4r_rma.h]: Remove assert?

File: src/mpid/ch4/src/ch4r_rma.h [hunk 12]
```diff
@@ -716,6 +717,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_CH4U_mpi_rget(void *origin_addr,
...
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4r_rma.h]: Remove assert?

File: src/mpid/ch4/generic/mpidig_send.h [hunk 13]
```diff
@@ -36,12 +36,15 @@ static inline int MPIDI_am_isend(const void *buf, MPI_Aint count, MPI_Datatype d
...
+            MPIR_ERR_CHKANDSTMT((*request) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail,
+                                "**nomemreq");
             MPID_Request_complete((*request));
         }
         goto fn_exit;
     }
 
     sreq = MPIDI_CH4I_am_request_create(MPIR_REQUEST_KIND__SEND, 2);
+    MPIR_ERR_CHKANDSTMT((sreq) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
     MPIR_Assert(sreq);
```
Comment (raffenet) [inline src/mpid/ch4/generic/mpidig_send.h]: Remove assert?

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
changes look good. just a few labels to add it looks like from the tests.

**Comment by yfguo** [comment]
Yep. Missed those when cleaning up the asserts

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
Looking at the am-only failures.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx
